### ISSUE TITLE
Swarm nodes cannot join as masters if http proxy is set

### ISSUE THREAD

**Issue by nealey** [moby/moby#36951]
#35395 has been closed but the issue persists in 17.12.1-ce **for masters only**, as evidenced by the multiple follow-ups on the closed bug. In case the team doesn't look at activity on closed bugs, I am opening a new one.

**Steps to reproduce the issue**

1. Setup Docker as usual. Add a proxy configuration like this to `/etc/systemd/system/docker.service.d/proxy.conf`:
```
[Service]
Environment="HTTP_PROXY=http://proxy.internal.company.com:8080" "HTTPS_PROXY=http://proxy.internal.company.com:8080" "NO_PROXY=localhost,127.0.0.1,.internal.company.com"
```
Don't forget to `systemctl daemon-reload` and restart the Docker daemon.

2. Initialize swarm on the master node
```
master# docker swarm init
Swarm initialized: current node (0tf92vjqqivr9mujdd99ip6ob) is now a manager.

To add a worker to this swarm, run the following command:

    docker swarm join --token SWMTKN-1-20n86wgp1jlfdku52x5pzlcwc6youavmzm78junqe7wtxqg4t5-2orbo6etiiknwjcfj8g52chtr 10.28.0.223:2377

To add a manager to this swarm, run 'docker swarm join-token manager' and follow the instructions.
```

3. Obtain token for manager
```
master# docker swarm join-token manager
    docker swarm join --token SWMTKN-1-5rouv1k0jt6cyf5gt2at2ai5vcw9a64j6sc8kwsk1dlkuwi4ej-0amahulewmk1e1y994uj8spm4 10.28.0.223:2377
```

4. Add a manager as indicated:

```
manager2# docker swarm join --token SWMTKN-1-5rouv1k0jt6cyf5gt2at2ai5vcw9a64j6sc8kwsk1dlkuwi4ej-0amahulewmk1e1y994uj8spm4 10.28.0.223:2377
```

Swarm join hangs because docker attempts to connect to the master through the proxy. Then it reports

    Error response from daemon: rpc error: code = Unavailable desc = grpc: the connection is unavailable

**Describe the results you received:**

see above

**Describe the results you expected:**

Masters should join without requiring addition of each machine's IP in the `no_proxy` environment variable passed to `dockerd`.

**Additional information you deem important (e.g. issue happens only occasionally):**

```
core@otter ~ $ docker version
Client:
 Version:       17.12.1-ce
 API version:   1.35
 Go version:    go1.9.4
 Git commit:    7390fc6
 Built: Tue Feb 27 22:10:31 2018
 OS/Arch:       linux/amd64

Server:
 Engine:
  Version:      17.12.1-ce
  API version:  1.35 (minimum version 1.12)
  Go version:   go1.9.4
  Git commit:   7390fc6
  Built:        Tue Feb 27 22:10:31 2018
  OS/Arch:      linux/amd64
  Experimental: true
```

**Comment by nealey** [comment]
Here is how to work around:

Edit a file `/etc/systemd/system/docker.service.d/proxy.conf` to list the IP address of every machine that might wish to become a swarm manager. Since we run several swarms and have central provisioning, this is a very large list for us:

```
[Service]
Environment=no_proxy=10.28.0.223,10.28.0.222,10.28.0.221,10.28.0.220,10.28.0.219,10.28.0.218,10.28.0.217,10.28.0.216,10.28.0.205,10.28.0.204,10.28.0.203,10.28.0.185,10.28.0.183,10.28.0.182,10.28.0.181,10.28.0.119,10.28.0.169,10.28.0.199,10.28.0.178,10.28.0.234,10.28.0.161,10.28.0.153,10.28.0.233,10.28.0.116,10.28.0.184,10.28.0.250,10.28.0.210,10.28.0.163,10.28.0.243,10.28.0.251,10.28.0.252,10.28.0.158,10.28.0.159,10.28.0.160,10.28.0.162,10.28.0.164,10.28.0.150,10.28.0.151,10.28.0.152,10.28.0.120,10.28.0.154,10.28.0.155,10.28.0.209,10.28.0.206,10.28.0.186,10.28.0.207,10.28.0.208,10.28.0.211,10.28.0.212,10.28.0.213,10.28.0.214,10.28.0.189,10.28.0.215,10.28.0.246,10.28.0.165,10.28.0.166,10.28.0.167,10.28.0.168,10.28.0.170,10.28.0.171,10.28.0.172,10.28.0.173,10.28.0.187,10.28.0.188,10.28.0.174,10.28.0.175,10.28.0.235,10.28.0.236,10.28.0.237,10.28.0.238,10.28.0.239,10.28.0.240,10.28.0.241,10.28.0.121,10.28.0.122,10.28.0.190,10.28.0.191,10.28.0.242,10.28.0.244,10.28.0.192,10.28.0.193,10.28.0.194,10.28.0.195,10.28.0.176,10.28.0.177,10.28.0.123,10.28.0.124,10.28.0.125,10.28.0.126,10.28.0.127,10.28.0.128,10.28.0.129,10.28.0.130,10.28.0.131,10.28.0.132,10.28.0.245,10.28.0.247,10.28.0.248,10.28.0.196,10.28.0.197,10.28.0.198,10.28.0.249,10.28.0.253,10.28.0.254,10.28.0.200,10.28.0.201,10.28.0.202,10.28.0.133,10.28.0.134,10.28.0.135,10.28.0.136,10.28.0.137,10.28.0.138,10.28.0.139,10.28.0.140,10.28.0.141,10.28.0.142,10.28.0.143,10.28.0.144,10.28.0.145,10.28.0.146,10.28.0.147,10.28.0.148,10.28.0.149,10.28.0.156,10.28.0.157,10.28.0.179,10.28.0.180
```

**Comment by nealey** [comment]
Something's changed and we can no longer get a manager joined under any condition. It may not be the proxy after all. We will spend the day attempting to locate helpful log information and system traces.

**Comment by nealey** [comment]
It appears that it is the proxy, but it's containerd that has the problem. If the docker team doesn't maintain containerd, can you direct me to who does, so I can file a bug there instead?

**Comment by thaJeztah** [comment]
curious: what distro are you on, and is this an official Docker package, or a docker package built by your distro? (As the official packages currently don't run containerd as a separate unit / service)

**Comment by thaJeztah** [comment]
Guess the situation would be the same if not started separarely; also Wondering how the proxy could influence dockerd not joining the cluster

**Comment by nealey** [comment]
My configuration sets http_proxy and https_proxy for every child of PID 1. Tcpdump showed a DNS request for the proxy hostname every time I tried to join, from the machine trying to join. Containerd wound up being the remaining thing needing a change, but I didn't investigate what it was trying to do. If it's also using the gRPC library, presumably it's affected by the same issue.

**Comment by kleptog** [comment]
@nealey no_proxy may work, but it also means you need to restart your entire swarm if you want to add a new master node. This is a non-starter in our case.

**Comment by nealey** [comment]
We added every IP in the subnet. Not an elegant solution, but it allowed us to move on to other problems.

**Comment by skrobul** [comment]
Just stumbled upon exact same issue. In my case we had the `NO_PROXY` configured correctly (at least according to documentation) but it still did not work. It turned out to be a problem with values provided as CIDR subnets rather than individual IPs.

Changing

```
NO_PROXY=localhost,127.0.0.0/8,10.0.0.0/8
```

to 
```
NO_PROXY=localhost,127.0.0.0/8,10.0.3.4,10.0.3.5,10.0.3.6,10.0.3.7,10.0.3.8,10.0.3.9
```

worked, but it's ugly workaround not a real solution as it defats whole idea of dynamic discovery. Also as @kleptog mentioned, restarting entire swarm just to add new manager node is crazy.

Observed on Docker 18.03.1-ce running on Debian hosts with 4.9.88 kernel.

**Comment by hgtok** [comment]
Facing same issue on CentOS 7. 
Have to set no_proxy to containerd.service to get managers to join successfully.

For those worried having to restart for new managers, you can plan ahead by whitelisting the whole subnet using this syntax. (Assuming managers are deployed on same subnet)

NO_PROXY=10.0.0.*

**Comment by thaJeztah** [comment]
This should be fixed in Docker 18.09.4 through https://github.com/docker/engine/pull/170
